source/slang/slang-emit-spirv.cpp

```diff
@@ -0,0 +1,1141 @@
+        /// This function should only be called once.
+        ///
+    void emitPhysicalLayout()
+    {
+        // [2.3: Physical Layout of a SPIR-V Module and Instruction]
+        //
+        // > Magic Number
+        //
+        m_words.add(SpvMagicNumber);
+
+        // > Version nuumber
+        //
+        m_words.add(SpvVersion);
+
+        // > Generator's magic number.
+        // > Its value does not affect any semantics, and is allowed to be 0.
+        //
+        // TODO: We should eventually register a non-zero
+        // magic number to represent Slang/slangc.
+        //
+        m_words.add(0);
+
+        // > Bound
+        //
+        // As described above, we use `m_nextID` to allocate
+        // <id>s, so its value when we are done emitting code
+        // can serve as the bound.
+        //
+        m_words.add(m_nextID);
+
+        // > 0 (Reserved for instruction schema, if needed.)
+        //
+        m_words.add(0);
+
+        // > First word of instruction stream
+        // > All remaining words are a linear sequence of instructions.
+        //
+        // Once we are done emitting the header, we emit all
+        // the instructions in our logical sections.
+        // 
+        for( int ii = 0; ii < int(SpvLogicalSectionID::Count); ++ii )
+        {
+            m_sections[ii].dumpTo(m_words);
+        }
+    }
+
+    // We will often need to refer to an instrcition by its
+    // <id>, given only the Slang IR instruction that represents
+    // it (e.g., when it is used as an operand of another
+    // instruction).
+    //
+    // To that end we will keep a map of instructions that
+    // have been emitted, where a Slang IR instruction maps
+    // to the corresponding SPIR-V instruction.
+
+        /// Map a Slang IR instruction to the corresponding SPIR-V instruction
+    Dictionary<IRInst*, SpvInst*> m_mapIRInstToSpvInst;
+
+        /// Register that `irInst` maps to `spvInst`
+    void registerInst(IRInst* irInst, SpvInst* spvInst)
+    {
+        m_mapIRInstToSpvInst.Add(irInst, spvInst);
+    }
+
+    // When we are emitting an instruction that can produce
+    // a result, we will allocate an <id> to it so that other
+    // instructions can refer to it.
+    //
+    // We will allocate <id>s on emand as they are needed.
+
+        /// Get the <id> for `inst`, or assign one if it doesn't have one yet
+    SpvWord getID(SpvInst* inst)
+    {
+        auto id = inst->id;
+        if( !id )
+        {
+            id = m_nextID++;
+            inst->id = id;
+        }
+        return id;
+    }
+
+    // We will build up `SpvInst`s in a stateful fashion,
+    // mostly for convenience. We could in theory compute
+    // the number of words each instruction needs, then allocate
+    // the words, then fill them in, but that would make the
+    // emit logic more complicated and we'd like to keep it simple
+    // until we are sure performance is an issue.
```

Comment:
There are several inter-related issues that make dynamically allocating the instructions kind of a local optimum, even if we could in principle avoid doing it completely.

For example, we have the dual complications that:

* Only a subset of the Slang IR instructions need to be emitted to SPIR-V (e.g., because sometimes there will be literals only used by decorations that SPIR-V doesn't care about)
* The Slang IR instructions at module scope don't currently enforce a dependency-based ordering, so naively emitting them in order would violate SPIR-V rules.

Those two birds are hit with the one stone of emitting SPIR-V instructions on-demand, but as soon as we emit thing on demand you have the problem that if you have `someInst(a, b, c, d, ...)` and you need to emit `a, b, c, ...` before `somInst`, then you can either:

* Start building `someInst` and appending the operands for `a, b, c, ...` to it one-by-one while emitting them on-demand such that they will eventually come before `someInst`. This is what the `SpvInst` mess lets us do.

* First ensure that all of `a, b, c, ...` have been emitted and collect their `<id>`s into an array, then emit the `someInst` using those operands.

That second option isn't really any better in practice because it still involves allocating scratch memory that is proportional in size to the instruction. The only way to not have to allocate that kind of scratch memory *at all* would be to deal with the original issues via IR passes:

* Eliminate all the decorations/instructions that aren't relevant to SPIR-V emit from the global scope of the module. This would be a kind of SPIR-V-specific dead code pass.

* Sort all the instructions at the global scope of the module to respect some or all of the SPIR-V ordering constraints. This would be a SPIR-V-specific transformation pass.

There's nothing stopping us from writing those passes, but at that point information about the SPIR-V lowering process has to be shared and coordinated across more places in the code (two or more passes need to agree about what is/isn't relevant to SPIR-V).

Reply:
I suspect that the easiest solution in practice would be to decouple the two concepts of:

* Emitting a SPIR-V instruction
* Allocating a chunk of "scratch" memory so that we can write words into it

If instead of `SpvInst` we just had a `SpvChunk` construct that held a sequence of words (pool-allocated), then when emitting an instruction would could choose on a case-by-case basis whether to:

* Pre-emit any required operands (in case they land in the same chunk), pre-compute the total number of words required, and then just emit the words of the instruction directly to the output chunk

* Allocate a fresh sub-chunk and fill it in piece-by-piece (knowing that nobody else can try to write to it), then append that sub-chunk to the destination.

I believe that such an approach is potentially a "best of both worlds" solution, especially if we can ensure that the common cases for instructions don't have to allocate anything.

Reply:
I think I lost my train of thought when writing some of these comments, so I'll note that one other (potentially weak) argument for having an intermediate `SpvInst` sort of structure is that we might want to ensure that the `<id>`s in our generated code are "canonical" or "minimal" in the sense of:

* Only giving a non-zero result `<id>` to instructions whose output value actually gets used

* Ensuring that the `<id>`s we generate are in some canonical ordering (e.g., `<id>` order exactly matches definition order, or order of first reference, or something similar)

This is realistically a very nit-picky thing to worry about, but if we ensure that our output is "canonical" in some way, then we can ensure that changes to the compiler implementation won't result in spurious diffs to generated SPIR-V (e.g., we start generating SPIR-V that is identical except for the `<id>`s and now tests start failing).

Any canonicalization we want to do could be done as a post-process on the SPIR-V, and re-numbering the `<id>`s could even be done in-place in memory, so this line of argument doesn't realistically hold water.